Add support for @mentioning users in hubot message

kdawgwilk opened the pull request:
Feature from #132 

Not sure if this is the right way to go about adding this feature but I did it for our company bot and figured others would like the feature as well.  I implemented in the send method so that you only need to prefix usernames with @ in your scripts and the when it goes through the adapter's send method the `@username` gets replaced with `<@UXXXXX>` to mention a user.  If the the user does not exist then the `@username` gets sent without actually mentioning or sending notification.  

FIXED:
Also I had a weird scoping issue with `@client` (I'm not super familiar with coffeescript) but solved it by making a copy (eg. `client = @client`) if anyone knows a better way to solve that I would be happy to change it

kdawgwilk commented:
#132

fcoelho commented on src/slack.coffee:
```diff
@@ -203,6 +205,14 @@ class SlackBot extends Adapter
       continue if msg.length < SlackBot.MIN_MESSAGE_LENGTH
 
+      # Replace @username with <@UXXXXX> for mentioning users
+      msg = msg.replace /(?:^@| @)([A-z]+)/gm, (match, p1) ->
```
you should be able to use a "fat arrow" in here to avoid the `client = @client` part

http://coffeescript.org/#fat-arrow

kdawgwilk commented:
Thanks @fcoelho got that fixed!

hkdobrev commented:
@kdawgwilk I saw your PR after I made #195. I was able to get everything working only after I've merged both. Your PR is replacing mentions in the `send` method and mine was modifying how the `reply` method was structuring the message. I think #195 now fixes everything. I've added your changes in a single commit. Please let me know what you think on #195. Cheers!

kdawgwilk commented:
@hkdobrev I added support for mentioning channel, everyone, and group now as well not sure if you want those changes on your branch as well

hkdobrev commented:
@kdawgwilk Great!

Since your PR is bigger and it's the more useful one I'd suggest to just include my change from https://github.com/hkdobrev/hubot-slack/commit/3fe59755628ffb743e2fcf74587cabf85083a024. It just improves `robot.reply` to mention with `@` in addition to your changes. If you do that I'll happily close my PR.

hkdobrev commented:
Any input from @paulhammond and @evansolomon? Thanks!

kdawgwilk commented:
@hkdobrev alright made that change and I am just going to add a few tests to make sure everything is looking good

kdawgwilk commented:
Also Sorry about the non-descriptive branch name everyone, I thought it was only going to be a few lines

kdawgwilk commented:
Just added tests and refactored a few things.  I have one failing test and cant seem to get to the bottom of it if anyone else wants to take a look at it that would be greatly appreciated.  I will keep looking at it as well.

kdawgwilk commented:
Sweet figured it out! Just was some issue with coffeescript I didn't understand

siminm commented:
Looking forward to seeing this merged!

jkimbo commented:
:+1: !

jtomaszon commented:
:+1: to merge!

aranasaurus commented:
I'd really love to see this merged! Or as an alternative, get pointed in the right direction for how to point my hubot at this branch for its slack adaptor?

kylestev commented:
LGTM! :+1:

aranasaurus commented:
Looks like there used to be a config option for this, but it was removed here: https://github.com/slackhq/hubot-slack/commit/c14b4d11cfd48dc547e5660e027991e85ba3ef1e and never returned.

It seems this used to be being done server-side based on an arg being passed to `/services/hooks/hubot`. Now the adapter is sending through `slack-client` and from what I can tell this may actually be as simple as passing an argument `link_names=1` through the `channel.send` as documented in the `Parsing Modes` section here: https://api.slack.com/docs/formatting.

So I _think_ all of this PR could be replaced with adding the `HUBOT_SLACK_LINK_NAMES` env variable in to `options` in `run` and just setting `message.link_names = options.link_names` before sending any messages.

I'll try doing this when I get a chance, but I've spent my time for the morning on it just spelunking, and wanted to share what I found in case somebody with the time came across this ;)

aranasaurus commented:
I finally got some time to put toward this again and since everything is going through the websocket api (via `client._send()` it's not as simple as setting link_names to true :(. Setting link_names to true does work when using methods in the adapter that send through `client._apiCall()` though.

So it seems like the way this PR does it is the right way to do it. Would really like to see this merged.

barbosa commented:
Any plans to merge this?

QuentinFchx commented:
@paulhammond could you merge this ? :)

tedsteen commented:
:+1:

QuentinFchx commented:
@mjsz sorry to insist but if you could merge this (or explain why it's not mergeable) it would be awesome :)

mjsz commented:
Looks reasonable, and I should be able to get this merged later this week. Thanks for your patience.

tedsteen commented:
We've been using this for quite some time and it's been working fine

mjsz commented on src/slack.coffee:
```diff
@@ -203,6 +203,16 @@ class SlackBot extends Adapter
+        user = @client.getUserByName p1
+        if user
+          match = match.replace /@[\w]+/, "<@#{user.id}>"
+        else if (p1 is 'channel' or p1 is 'everyone' or p1 is 'group')
```
It'd be nice to pull out the array @ line 181 and then use that here for a membership check.

(Also be sure to sync to get the merge of https://github.com/slackhq/hubot-slack/pull/218)

mjsz commented:
merged via https://github.com/slackhq/hubot-slack/pull/246
